coil/src/main/java/dev/chrisbanes/accompanist/coil/Coil.kt

```diff
@@ -135,65 +122,72 @@ fun CoilImage(
     loading: @Composable (() -> Unit)? = null,
     onRequestCompleted: (RequestResult) -> Unit = emptySuccessLambda
 ) {
-    WithConstraints(modifier) {
-        val requestWidth = constraints.requestWidth.value
-        val requestHeight = constraints.requestHeight.value
+    var requestSize by state { IntPxSize.Zero }
+    var result by state<RequestResult?> { null }
+
+    launchInComposition(requestSize) {
+        val requestWidth = requestSize.width.value
+        val requestHeight = requestSize.height.value
 
-        // Execute the request using executeAsComposable(), which guards the actual execution
-        // so that the request is only run if the request changes.
-        val result = when {
+        result = when {
             request.sizeResolver != null -> {
                 // If the request has a sizeResolver set, we just execute the request as-is
-                request.executeAsComposable()
+                request.execute()
             }
             requestWidth > 0 && requestHeight > 0 -> {
                 // If we have a non-zero size, we can modify the request to include the size
                 request.newBuilder()
                     .size(requestWidth, requestHeight)
                     .build()
-                    .executeAsComposable()
-            }
-            else -> {
-                // Otherwise we have a zero size, so no point executing a request
-                null
+                    .execute()
             }
+            // Otherwise we have a zero size, so no point executing a request
+            else -> null
         }
+    }
 
-        onCommit(result) {
-            if (result != null) {
-                onRequestCompleted(result)
-            }
-        }
+    onCommit(result) {
+        // Execute the onRequestCompleted callback if we have a new result
+        result?.also { onRequestCompleted(it) }
+    }
 
-        val painter = when (result) {
-            is SuccessResult -> {
-                if (getSuccessPainter != null) {
-                    getSuccessPainter(result)
-                } else {
-                    defaultSuccessPainterGetter(result)
-                }
+    val painter = when (val r = result) {
```

Comment:
Try to avoid doing this work in composition

Reply:
Yeah, this is necessary currently because `getSuccessPainter` is `@Composable`. I'll remove this eventually once I think of a good way to implement transitions.